### PR TITLE
fix(router) split paths before marshalling routes

### LINKED ISSUE #5438: Longest path routing with multiple hosts and the same path is broken
### Summary

After #5418 is fixed, longest path routing still breaks when I add more than one route with the same path and different host to my declarative configuration piece. Wrong service is getting matched.

### Steps To Reproduce

1. Start kong in db-less mode with the following kong.conf and kong.yaml:
```
docker run -p 8000:8000 -p 8001:8001 -v `pwd`/kong.yaml:/config/kong.yaml -v `pwd`/kong.conf:/config/kong.conf kong:1.3.0 kong start -c /config/kong.conf
```

kong.conf:
```
database = off
declarative_config = /config/kong.yaml
proxy_access_log = /dev/stdout
admin_listen = 0.0.0.0:8001
admin_access_log = /dev/stdout
```

kong.yaml:
```
_format_version: "1.1"

services:
  - name: devportal-public
    host: localhost
    port: 8001
    path: /routes
    routes:
      - name: public-apiv1
        hosts:
          - api.local
        paths:
          - /rest/devportal/api/v1
          - /rest/devportal

  - name: devportal-aux
    host: localhost
    port: 8001
    path: /services
    routes:
      - name: aux
        hosts:
          - api.local
        paths:
          - /rest/devportal/aux

  - name: devportal-aux-host
    host: localhost
    port: 8001
    path: /plugins
    routes:
      - name: aux-host
        hosts:
          - test-api.local
        paths:
          - /rest/devportal/aux

  - name: devportal-aux-host2
    host: localhost
    port: 8001
    path: /endpoints
    routes:
      - name: aux-host2
        hosts:
          - atest-api.local
        paths:
          - /rest/devportal/aux
```
Added devportal-aux-host2 service and route compared to kong.yaml in #5418.

2. Call kong as follows:
```
curl -H 'Host:test-api.local' localhost:8000/rest/devportal/aux
## Edit: curl -H 'Host:api.local' localhost:8000/rest/devportal/aux
```
With kong-nightly/master/2020-01-11 public-apiv1 is getting matched. I expect aux route match.

### Additional Details & Logs

kong-nightly/master 2020-01-11 (includes fix for #5418)

### PR DESCRIPTION
**Pull request by locao**
Paths must be evaluated independently when sorting for route matching.

Fix #5438

### CODE REVIEW

File: kong/router.lua [hunk 1]
```diff
@@ -1046,21 +1046,32 @@ function _M.new(routes)
...
 
     for i = 1, #routes do
-      local route_t, err = marshall_route(routes[i])
-      if not route_t then
-        return nil, err
-      end
 
-      if route_t.route.paths ~= nil and #route_t.route.paths > 1 then
+      if routes[i].route.paths ~= nil and #routes[i].route.paths > 1 then
+        local paths = routes[i].route.paths
```
Comment (bungle) [inline kong/router.lua]: I would write this as:
```lua
local paths = routes[i].route.paths
if paths ~= nil and #paths > 1 then     
```

**Comment by bungle** [comment]
I tried this, and it does fix the #5438, good job!

**Comment by atikhono** [comment]
@bungle I tried to test this patch as well, works with the declarative config in the issue description. But as long as I another simple configuration block:
```
  - name: devportal2
    host: localhost
    port: 8001
    path: /routes
    routes:
      - name: devportal-route-2
        hosts:
          - atest-api.local
        paths:
          - /rest/devportal
```
I get 404 "no Route matched with those values" for `curl -H 'Host:atest-api.local' localhost:8000/rest/devportal` (and I expect routing to Kong admin).

These blocks are building blocks of my 450-line declarative configuration file that has been working in production with Kong 1.2.2. I'm desperately trying to migrate to Kong 1.4+ or 2.0 when released (highly anticipated!) :).

I have tried to run my original and full configuration file with the PR changes, and still getting various issues like bad route matching somewhere else and even parsing errors (like `parse error: Invalid numeric literal at line 1, column 7`). Though I wasn't able to build kong properly using this article: https://docs.konghq.com/install/source/. I ended up building a deb package from https://github.com/Kong/kong-build-tools (`make actual-package-kong`) and install it in docker using ubuntu Dockerfile from https://github.com/Kong/docker-kong based on Bionic. Still kong complains in missing openresty patches... So I cannot really tell test to report issues faster without a nightly build. Pew, just some grouching.

**Comment by hishamhm** [comment]
@atikhono Hi, I could not reproduce this second issue. Did you mean adding the above block to the config described in #5438? This is the yaml I'm testing with:

```yaml
_format_version: "1.1"

services:
  - name: devportal-public
    host: localhost
    port: 8001
    path: /routes
    routes:
      - name: public-apiv1
        hosts:
          - api.local
        paths:
          - /rest/devportal/api/v1
          - /rest/devportal

  - name: devportal-aux
    host: localhost
    port: 8001
    path: /services
    routes:
      - name: aux
        hosts:
          - api.local
        paths:
          - /rest/devportal/aux

  - name: devportal-aux-host
    host: localhost
    port: 8001
    path: /plugins
    routes:
      - name: aux-host
        hosts:
          - test-api.local
        paths:
          - /rest/devportal/aux

  - name: devportal-aux-host2
    host: localhost
    port: 8001
    path: /endpoints
    routes:
      - name: aux-host2
        hosts:
          - atest-api.local
        paths:
          - /rest/devportal/aux

  - name: devportal2
    host: localhost
    port: 8001
    path: /routes
    routes:
      - name: devportal-route-2
        hosts:
          - atest-api.local
        paths:
          - /rest/devportal
```

Then I'm running this:

```
curl -H 'Host:atest-api.local' localhost:8000/rest/devportal
```

and I get the /routes as expected, no 404 here:
```
{"next":null,"data":[{"id":"0ea7d854-22a8-5f82-bad2-cf8690d3ac43","tags":null,"updated_at":1579293543,"destinations":null,"headers":null,"protocols":["http","https"],"created_at":1579293543,"snis":null,"hosts":["atest-api.local"],"name":"aux-host2","preserve_host":false,"regex_priority":0,"strip_path":true,"sources":null,"paths":["\/rest\/devportal\/aux"],"https_redirect_status_code":426,"service":{"id":"beb03005-5da2-5467-9904-e62757138541"},"methods":null},{"id":"f8e3c07a-42ef-5d4c-8d34-53e534497e3c","tags":null,"updated_at":1579293543,"destinations":null,"headers":null,"protocols":["http","https"],"created_at":1579293543,"snis":null,"hosts":["api.local"],"name":"aux","preserve_host":false,"regex_priority":0,"strip_path":true,"sources":null,"paths":["\/rest\/devportal\/aux"],"https_redirect_status_code":426,"service":{"id":"48f67d99-e139-5f81-a140-b8ed57bd38ff"},"methods":null},{"id":"ca57d3f9-3e23-5ac6-9e94-553bdf8e350c","tags":null,"updated_at":1579293543,"destinations":null,"headers":null,"protocols":["http","https"],"created_at":1579293543,"snis":null,"hosts":["test-api.local"],"name":"aux-host","preserve_host":false,"regex_priority":0,"strip_path":true,"sources":null,"paths":["\/rest\/devportal\/aux"],"https_redirect_status_code":426,"service":{"id":"ced472fc-a050-5c2f-8224-79bdcfbd1b2c"},"methods":null},{"id":"db6b60e3-810a-507a-8842-880b678ef470","tags":null,"updated_at":1579293543,"destinations":null,"headers":null,"protocols":["http","https"],"created_at":1579293543,"snis":null,"hosts":["atest-api.local"],"name":"devportal-route-2","preserve_host":false,"regex_priority":0,"strip_path":true,"sources":null,"paths":["\/rest\/devportal"],"https_redirect_status_code":426,"service":{"id":"b05dc5df-ca9b-5781-9e44-3796a29be97f"},"methods":null},{"id":"a6ba5120-0f81-57a8-abc9-cc4873cb5485","tags":null,"updated_at":1579293543,"destinations":null,"headers":null,"protocols":["http","https"],"created_at":1579293543,"snis":null,"hosts":["api.local"],"name":"public-apiv1","preserve_host":false,"regex_priority":0,"strip_path":true,"sources":null,"paths":["\/rest\/devportal\/api\/v1","\/rest\/devportal"],"https_redirect_status_code":426,"service":{"id":"2672ed5c-a244-5dbe-95b1-44ed24c9e808"},"methods":null}]}hisham@proxy
```

Is this the scenario that failed for you with this patch or did you have something else in mind?

**Comment by atikhono** [comment]
@hishamhm Good to know, let's have it merged then (when concerns addressed)! Looking forward to test a nightly build ASAP. I'm getting a stable 404 with that config calling atest-api.local/rest/devportal. Hopefully, it's my "bad" build even though I don't see how it could intervene.

**Comment by hishamhm** [comment]
@atikhono Sounds like a plan! Thank you for all your feedback on this, and we'll keep an eye if we spot anything else on this department. I believe this is the last fix we are going to merge before 2.0.0, so please give the final release a spin and let us know your findings!

**Comment by marckhouzam** [comment]
Thanks for fixing this bug!

We hit it in our first migration from Kong 0.14 to 1.4.  Will there be a 1.4.4 release to include this fix?

**Comment by hishamhm** [comment]
@marckhouzam We recommend migrating to 1.5.0 instead. It includes this fix and provides better migrations from 0.14.1!

**Comment by marckhouzam** [comment]
Thanks for the quick reply @hishamhm 

> @marckhouzam We recommend migrating to 1.5.0 instead. It includes this fix and provides better migrations from 0.14.1!

We have already migrated our Development environment to 1.4.2.  If I migrate it again from 1.4.2 to 1.5 (so I can soak it before upgrading production to 1.5), I assume I will get all `"path_handling"` set to `"v1"`.  To properly test what will happen in production, I gather I should manually change all `"path_handling"` to be `"v0"`.  Does that make sense?

**Comment by atikhono** [comment]
> @atikhono Sounds like a plan! Thank you for all your feedback on this, and we'll keep an eye if we spot anything else on this department. I believe this is the last fix we are going to merge before 2.0.0, so please give the final release a spin and let us know your findings!

All I can say so far 1.5.0 worked like a charm. Thanks to Kong team for fixing the issues quickly, and getting them released too. Much appreciated!

**Pull request merged**